Add per-model-instance input/output ports to MultibodyPlant

sammy-tri opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/9041)
<!-- Reviewable:end -->

sammy-tri commented:
+@rcory as a followup to #8986.  cc @amcastro-tri 

---

Review status: all discussions resolved, platform LGTM missing

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041#-:-LFYJQHXEzCFnL4g3ZHm:b-j88ja3)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
This is so great! just a few super high level comments.

---

Review status: 5 unresolved discussions, LGTM missing from assignee rcory, platform LGTM missing

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 284 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYN8b27_qdcWyd0FVA:-LFYN8b27_qdcWyd0FVB:b-64nwsw) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L284)):*
> ```objc
>   ///   part of the model. See HasBodyNamed(), Body::name().
>   /// @param[in] model_instance
>   ///   A model instance index which this body is part of.
> ```

nit: explain how to obtain one for those who create models programmatically?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 312 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYNNtbDHUMYu3Ti4md:-LFYNNtbDHUMYu3Ti4me:b-i8lrs3) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L312)):*
> ```objc
>   /// This method returns a constant reference to the body just added, which
>   /// will remain valid for the lifetime of `this` %MultibodyPlant.  The body
>   /// will use the default model instance.
> ```

nit: could you mention how to obtain the default model instance index?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 487 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYNWIa960qqDPmi7Wc:-LFYNWIa960qqDPmi7Wd:b-1jbgxc) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L487)):*
> ```objc
> 
>   /// Creates a new model instance.  Returns the index for the model
>   /// instance (as there is no concrete object beyond the index.
> ```

Parenthesis not closed. Also, difficult to parse (at least for me) what you meant with the last bit after the parenthesis. 

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1419 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYOP2j5rf04Om2xMsf:-LFYOP2j5rf04Om2xMsg:b-1qwzpo) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1419)):*
> If only the default model instance is being used, this vector will be empty.

couldn't my entire robot, which presumably could be actuated, belong to the default instance?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 16 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYeSfRD_dcnr8vGVbT:-LFYeSfRD_dcnr8vGVbU:biswprt) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L16)):*
> ```objc
> /// adds it to `plant`. The SDF file can only contain a single `<model>`
> /// element. `<world>` elements (used for instance to specify gravity) are
> /// ignored by this method.  A new model instance will be added to @p plant.
> ```

nit: Maybe `@ref` to the appropriate section in the MBP docs? (with an `@anchor`)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041#-:-LFYejRHBxHc93K_W1yj:b5cnwin)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all discussions resolved, LGTM missing from assignee rcory, platform LGTM missing

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 16 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYeSfRD_dcnr8vGVbT:-LFYuO3e-KTr7-G0jOhE:brk6upu) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L16)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I'm not sure what to link to...  `MultibodyPlant::AddModelInstance()` maybe?  Or just point to `model_instance.h`?
</blockquote></details>

Ah!, I see. Maybe if you place `AddModelInstance()`, `HasModelInstance()`, `GetModelInstanceIndexByName()` within a Doxygen section with an anchor you can say more about model instances and reference that?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041)*
<!-- Sent from Reviewable.io -->

rcory commented:
Very nice! A few comments. PTAL.

---

Reviewed 5 of 5 files at r1.
Review status: 1 unresolved discussion, LGTM missing from assignee rcory, platform LGTM missing

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 312 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYNNtbDHUMYu3Ti4md:-LFYiKmuAAKT1pcDOzhi:bw01pwr) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L312)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

nit: could you mention how to obtain the default model instance index?
</blockquote></details>

BTW, since you don't need to provide the default model instance to this function, perhaps an example for obtaining it won't make sense here?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 331 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYj2VW0JcB2ia5r1lv:-LFYj2VW0JcB2ia5r1lw:b-gffr4s) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L331)):*
> ```objc
>   ///   frame B.
>   /// @returns A constant reference to the new RigidBody just added, which will
>   ///          remain valid for the lifetime of `this` %MultibodyPlant.
> ```

Shouldn't we throw if more than the default and world model instances exist? (Like we do in RBT).

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 826 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYjwqN9jKfqx43Hvv9:-LFYjwqN9jKfqx43HvvA:bp1mjjx) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L826)):*
> ```objc
> 
>   /// Returns a constant reference to the input port for external actuation for
>   /// the case where only one model instance has actuated dofs..  This input
> ```

BTW extra period.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 839 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYkJPI7BzW-B58SCcJ:-LFYkJPJANj1rs65U6GL:b-gpe0vv) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L839)):*
> ```objc
>   /// @pre Finalize() was already called on `this` plant.
>   /// @throws if called before Finalize() or if the model instance does not
>   /// contain any actuators.
> ```

BTW what about throwing if the model instance does not exist?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 861 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYkY14Eu9UMmsmTarU:-LFYkY14Eu9UMmsmTarV:b-n6bxr4) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L861)):*
> ```objc
>   /// @pre Finalize() was already called on `this` plant.
>   /// @throws if called before Finalize() or if the model instance does not
>   /// have any state.
> ```

BTW throw if the model instance does not exist?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1419 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYOP2j5rf04Om2xMsf:-LFYlutc4hN5KiRvEMI9:bxd2drp) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1419)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Removed stale/misleading part of the comment (it referred to an implementation detail which was no longer accurate by the time I opened the PR).
</blockquote></details>

BTW would it be appropriate to assume that the size of this vector is the same as the number of available model instances?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1427 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYm72O3XHWKr5B1zSN:-LFYm72O3XHWKr5B1zSO:bae01fx) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1427)):*
> ```objc
>   int continuous_state_output_port_{-1};
>   // A vector containing state output ports for each model instance.  A value of
>   // -1 indicates that the model instance has no state.
> ```

BTW same question as above size of this vector =? valid instances?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 30 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYvGm9AOKwyCx6kwDO:-LFYvGm9AOKwyCx6kwDP:b-ve507y) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L30)):*
> ```objc
> /// @param model_name
> ///   The name given to the newly created instance of this model.  If
> ///   empty, the "name" attribute from the model tag will be used.
> ```

BTW name should be unique and throw if not?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 45 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYv_2q8iUoDX0O3zCL:-LFYv_2q8iUoDX0O3zCM:b-cx1n3l) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L45)):*
> ```objc
> 
> /// Alternate version of AddModelFromSdfFile which always uses the "name"
> /// element from the model tag for the name of the newly created model instance.
> ```

BTW name should be unique and throws if not?

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 296 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYwh9o0aTaCXe-zYWd:-LFYwh9o0aTaCXe-zYWe:bmjw1uj) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L296)):*
> ```cpp
>   EXPECT_EQ(plant_.GetModelInstanceByName("acrobot"), instance2_idx);
> }
> ```

BTW perhaps a test for unique model instance names?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041#-:-LFYwvYy2GWlRUwgfDYT:b-wc060f)*
<!-- Sent from Reviewable.io -->

rcory commented:
Reviewed 3 of 3 files at r2.
Review status: all discussions resolved, LGTM missing from assignee rcory, platform LGTM missing

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1427 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYm72O3XHWKr5B1zSN:-LFZaD4bAcP-yZGFPy_G:bj3taze) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1427)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Not necessarily, ports won't be allocated for empty model instances or ones with no continuous state (e.g. all links are welded, like a table or something)
</blockquote></details>

A little confused. Wouldn't you have a -1 element in the vector for these cases (and hence the size of the vector would still reflect the presence of these)?

---

*[multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h, line 30 at r1](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFYvGm9AOKwyCx6kwDO:-LFZZOXL4AkzG4C1MDfg:bjfrvqj) ([raw file](https://github.com/robotlocomotion/drake/blob/a3d8da9f3eb21b75c67153d8672f2c7f95cac831/multibody/multibody_tree/parsing/multibody_plant_sdf_parser.h#L30)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

MBT will throw eventually.  (also below).  Pondering on how much I feel like duplicating the same sanity check all of the way up the stack.  Added a test.
</blockquote></details>

Ah right! All good.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform pass done -- code looks beautiful! A few minor things below, ptal. Also, I'm not seeing unit tests for the new (and rather tricky) model instance methods for extracting and assembling model-instance-specific quantities. Are those around somewhere? If not, please add.

---

Reviewed 2 of 5 files at r1, 3 of 3 files at r2.
Review status: 13 unresolved discussions, LGTM missing from assignee sherm1, platform LGTM missing

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 310 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFchpUs8jupfmpgouHx:-LFchpUs8jupfmpgouHy:b-a3990n) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L310)):*
> ```objc
>   }
> 
>   /// Creates a rigid body model with the provided name and spatial inertia.
> ```

Nit: consider just "Creates a rigid body with ...". "Rigid body model" usually refers to a whole multibody system comprised of rigid bodies.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 313 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFciaUyEZFYqO0tDwb0:-LFciaUyEZFYqO0tDwb1:bej4ogp) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L313)):*
> ```objc
>   /// This method returns a constant reference to the body just added, which
>   /// will remain valid for the lifetime of `this` %MultibodyPlant.  The body
>   /// will use the default model instance (see @ref model_instance.h).
> ```

FYI referencing a header file is likely awkward for a Doxygen user. Is there a class or alias or @anchor that would make more sense here?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 594 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcjHqjDipBLRUQI4dC:-LFcjHqjDipBLRUQI4dD:bouet0t) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L594)):*
> ```objc
> 
>   /// Returns the index to the model instance that is uniquely identified
>   /// by the string `name` in `this` model.
> ```

Nit: model -> MultibodyPlant ?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 597 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcjQ5q21W6-3q1ILog:-LFcjQ5q21W6-3q1ILoh:bouet0t) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L597)):*
> ```objc
>   /// @throws std::logic_error if there is no instance with the requested name.
>   /// @see HasModelInstanceNamed() to query if there exists an instance in
>   /// `this` model with a given specified name.
> ```

Nit: model -> MultibodyPlant ?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 599 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcj_aOAgsrGQrOY2cT:-LFcj_aOAgsrGQrOY2cU:b-j21moe) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L599)):*
> ```objc
>   /// `this` model with a given specified name.
>   ModelInstanceIndex GetModelInstanceByName(const std::string& name) const {
>     return model_->GetModelInstanceByName(name);
> ```

Nit: I wonder if this data member is now misnamed since it contains multiple model instances? It seems misleading and it leaked into user comments above. The term is _so_ generic! How about `multibody_tree_` instead?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 853 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFclHaA0dUzx4D3dJDK:-LFclHaA0dUzx4D3dJDL:blg4g98) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L853)):*
> ```objc
>       ModelInstanceIndex model_instance) const;
> 
>   /// }@
> ```

Minor: should be `@}`. Please generate the doxygen to make sure it comes out as you wanted it.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 875 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFclgXZ9jh0lnAEbPJa:-LFclgXZ9jh0lnAEbPJb:b-vcd1h7) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L875)):*
> ```objc
>   const systems::OutputPort<T>& get_continuous_state_output_port(
>       ModelInstanceIndex model_instance) const;
>   /// }@
> ```

Minor: should be `@}`.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1125 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcm6fp9ebAHNilOjcB:-LFcm6fp9ebAHNilOjcC:bcvngkl) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1125)):*
> ```objc
>   // Helper method to assemble actuation input vector from the appropriate
>   // ports.
>   VectorX<T> EvaluateActuationInput(
> ```

Nit: consider "EvaluateAndAssembleActuationInput()" -- we use "Eval" a lot to mean just evaluate and return with no processing.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1297 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFckLicDF4o-pEPm9ix:-LFckLicDF4o-pEPm9iy:bpntq0e) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1297)):*
> ```objc
> 
>   // The entire multibody model.
>   std::unique_ptr<drake::multibody::MultibodyTree<T>> model_;
> ```

Nit: maybe rename this member something more specific, like `multibody_tree_`?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1430 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcmqfY86DpOx47WQO4:-LFcmqfY86DpOx47WQO5:b-nvcjnv) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1430)):*
> ```objc
>   // A vector containing actuation ports for each model instance.  A value of
>   // -1 indicates that the model instance has no actuated dofs.
>   std::vector<int> instance_actuation_ports_;
> ```

Minor: I had trouble understanding this. Consider saying explicitly that it should be indexed by ModelInstanceIndex. Also, what is the 'int' here? Is it an InputPortIndex or a count? If it is an InputPortIndex consider making this an `std::vector<InputPortIndex>` and using the index `is_valid()` method instead of the -1 as the "not actuated" flag.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1433 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcnUd6EKXR0jU2xQjg:-LFcnUd6EKXR0jU2xQjh:b-y7ctsd) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1433)):*
> ```objc
> 
>   // If only one model instance has actuated dofs, remember it here.
>   ModelInstanceIndex actuated_instance_;
> ```

Nit: is this invalid if there is more than one actuated instance?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1438 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcngkCE7ue6g4jNAXH:-LFcngkCE7ue6g4jNAXI:b-xo8vyg) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1438)):*
> ```objc
>   // A vector containing state output ports for each model instance.  A value of
>   // -1 indicates that the model instance has no state.
>   std::vector<int> instance_continuous_state_output_ports_;
> ```

Minor: same confusion here. Should this be indexed by ModelInstanceIndex and is the return a vector of OutputPortIndex ? Can you clarify?

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 288 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcgbKK2fjCR6tZmn0J:-LFcgbKK2fjCR6tZmn0K:bb2h9f2) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L288)):*
> ```cpp
>   const std::string acrobot_sdf_name = FindResourceOrThrow(
>       "drake/multibody/benchmarks/acrobot/acrobot.sdf");
>   int instance2_idx = AddModelFromSdfFile(acrobot_sdf_name, &plant_);
> ```

Nit: should be a ModelInstanceIndex rather than an int

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 295 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcgmDv5DOMkowdeVoA:-LFcgmDv5DOMkowdeVoB:b9dp2wq) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L295)):*
> ```cpp
>   // Check that a duplicate model names are not allowed.
>   EXPECT_THROW(AddModelFromSdfFile(full_name_, "instance1", &plant_),
>                std::logic_error);
> ```

Minor: consider using DRAKE_EXPECT_THROWS_MESSAGE() to check that the right message is being thrown.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041#-:-LFcq-LwDifWyJUsydYG:b-n3kmt8)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
https://github.com/RobotLocomotion/drake/blob/master/multibody/multibody_tree/test/model_instance_test.cc is the unit test for manipulating model instance state vectors <-> full tree state vectors.  Is that what you were asking about?

---

Review status: 2 unresolved discussions, LGTM missing from assignee sherm1, platform LGTM missing

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 310 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFchpUs8jupfmpgouHx:-LFcvD4W7DlJELy1b48V:b-1pf9kw) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L310)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Nit: consider just "Creates a rigid body with ...". "Rigid body model" usually refers to a whole multibody system comprised of rigid bodies.
</blockquote></details>

Done (on `MultibodyTree` too)

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 313 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFciaUyEZFYqO0tDwb0:-LFcwOvl6-MhY5Xko6sU:byyg1me) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L313)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI referencing a header file is likely awkward for a Doxygen user. Is there a class or alias or @anchor that would make more sense here?
</blockquote></details>

Done.  (the target of that reference was the @file comment for that header, but  I think it looks a little cleaner this way)

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 594 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcjHqjDipBLRUQI4dC:-LFcwk4A-RNyWuTqYnw1:b-88pdtl) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L594)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Nit: model -> MultibodyPlant ?
</blockquote></details>

Should this be a pervasive change?  The other GetFooByName and HasFooNamed messages use the same language.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 597 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcjQ5q21W6-3q1ILog:-LFd0_UF5voFnMBApN6B:bhd7n0i) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L597)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Nit: model -> MultibodyPlant ?
</blockquote></details>

Will resolve as above.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 599 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcj_aOAgsrGQrOY2cT:-LFcx-Pd8Cb0Fu-M_tZ4:bekqv6t) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L599)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Nit: I wonder if this data member is now misnamed since it contains multiple model instances? It seems misleading and it leaked into user comments above. The term is _so_ generic! How about `multibody_tree_` instead?
</blockquote></details>

Or just `tree_`?   @amcastro-tri what do you think?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 853 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFclHaA0dUzx4D3dJDK:-LFcxr2k6rTgx4cebJyc:b2yb1q7) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L853)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Minor: should be `@}`. Please generate the doxygen to make sure it comes out as you wanted it.
</blockquote></details>

Done.  (it *almost* looked right before if you didn't pay too much attention...)

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1297 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFckLicDF4o-pEPm9ix:-LFcyExYA4XFCK4Tw48B:b16jsue) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1297)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Nit: maybe rename this member something more specific, like `multibody_tree_`?
</blockquote></details>

Will resolve in discussion above on the same topic.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1430 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcmqfY86DpOx47WQO4:-LFcyVej2OA7L3TTBELK:b47v9p0) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1430)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Minor: I had trouble understanding this. Consider saying explicitly that it should be indexed by ModelInstanceIndex. Also, what is the 'int' here? Is it an InputPortIndex or a count? If it is an InputPortIndex consider making this an `std::vector<InputPortIndex>` and using the index `is_valid()` method instead of the -1 as the "not actuated" flag.
</blockquote></details>

Done.  It *is* an `InputPortIndex`! I somehow forgot that.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1433 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcnUd6EKXR0jU2xQjg:-LFczHSFEtkBO8yFvPZq:bo7uu3v) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1433)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Nit: is this invalid if there is more than one actuated instance?
</blockquote></details>

Done.  (added comment)

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 1438 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcngkCE7ue6g4jNAXH:-LFczy9y6uuqccl-AkuS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L1438)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Minor: same confusion here. Should this be indexed by ModelInstanceIndex and is the return a vector of OutputPortIndex ? Can you clarify?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041#-:-LFctW8h84Y7dHJ-gPl3:br6nnql)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Regarding unit tests -- what I meant was the new MultibodyPlant APIs that involve model instances. Likely the unit tests should be in multibody_plant_test.cc but I didn't see them there. I mean like num_position(instance), AddModelInstance(), get_continuous_state_output_port(instance), etc. Are those included in some other unit tests?

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending resolution/addition of unit tests for new APIs.

---

Review status: 2 unresolved discussions, platform LGTM from [sherm1]

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041#-:-LFd6RLZ8rYoxP-BszCw:b-3mapzc)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@sammy-tri: CI failures in Debug:
> terminate called after throwing an instance of 'std::runtime_error'
  what():  Explicitly constructing an invalid index. Type "drake::TypeSafeIndex<drake::systems::InputPortTag>", has an invalid value; it must lie in the range [0, 2³¹ - 1].

I _think_ this comes from copying an invalid index -- see below for possible fix.

---

Review status: 3 unresolved discussions, platform LGTM from [sherm1]

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 957 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFdMqm82Rm-AnjXAogE:-LFdMqm82Rm-AnjXAogF:bbu9evn) ([raw file](https://github.com/robotlocomotion/drake/blob/1956c90595321e7e3441bb9fa1ce4aec7d3eef91/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L957)):*
> ```cpp
>     if (instance_num_dofs == 0) {
>       // Add an invalid index.
>       instance_actuation_ports_.push_back(InputPortIndex());
> ```

Minor: I believe this is not allowed since it copies an invalid index. I believe the following will fix it:
```c++
instance_actuation_ports_.emplace_back();
```
since then it is only default-constructing and not copying.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041#-:-LFdMX6c3eOkgdiPav_J:b-b1w2eg)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Doh, my fault for not running a debug build before checking in.  Added unit tests for the shape of the tree and ports.  Didn't add a test for the actual demultiplexing of the state since I feel like the `model_instance_test` tests cover that.

---

Review status: 2 unresolved discussions, platform LGTM from [sherm1]

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 594 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcjHqjDipBLRUQI4dC:-LFd9Y0kB5147YhtgtEQ:b2q1fw7) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L594)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I think so -- thanks!
</blockquote></details>

Done.  (though I won't claim to have gotten all uses of the word "model" to describe the entire tree/plant).

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 599 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcj_aOAgsrGQrOY2cT:-LFdUjKf6cvF0IwMrjUE:bc4mw1r) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L599)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

`tree_` sounds good to me too and won't make lines go over 80 if renamed!
</blockquote></details>

After talking to @amcastro-tri we were a bit concerned that the name change would intermingle a lot of lines of name change with the functionality changes (and I thought it would bloat the PR a bit).  How would you feel about renaming `model_` (and by extension the `model()` accessor) in a future standalone PR?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 957 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFdMqm82Rm-AnjXAogE:-LFdUCkN7CIsCznFpI2g:bgmm1fl) ([raw file](https://github.com/robotlocomotion/drake/blob/1956c90595321e7e3441bb9fa1ce4aec7d3eef91/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L957)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Minor: I believe this is not allowed since it copies an invalid index. I believe the following will fix it:
```c++
instance_actuation_ports_.emplace_back();
```
since then it is only default-constructing and not copying.
</blockquote></details>

I don't think this was actually causing the assertion failure (I think it was because I left in the explicit -1 initializers for the geometry ports), but I've changed it to not use `push_back` anyway.

---

*[multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc, line 299 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFd2Pp45XSsTc880qH_:-LFd7Ex4A7T9Oiu3F5Pu:bo3pyk9) ([raw file](https://github.com/robotlocomotion/drake/blob/1956c90595321e7e3441bb9fa1ce4aec7d3eef91/multibody/multibody_tree/parsing/test/multibody_plant_sdf_parser_test.cc#L299)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI (just in case you didn't notice -- fine as is) this can also be a regular expression if you want to be less specific (like `.*foo.*bar.*`).
</blockquote></details>

I had not noticed!  I think I'm OK with this version for now, but good to know for the future.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041#-:-LFdURDvB905ELc6sguC:b-efxq6o)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Perfect -- thanks, @sammy-tri !

---

Reviewed 3 of 3 files at r4.
Review status: 1 unresolved discussion, platform LGTM from [sherm1]

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 599 at r2](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFcj_aOAgsrGQrOY2cT:-LFdY5904UzVm150yf8f:b-87kpmc) ([raw file](https://github.com/robotlocomotion/drake/blob/3aab00e5c671dc7e113b1d6d2ebf265bb7866292/multibody/multibody_tree/multibody_plant/multibody_plant.h#L599)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

After talking to @amcastro-tri we were a bit concerned that the name change would intermingle a lot of lines of name change with the functionality changes (and I thought it would bloat the PR a bit).  How would you feel about renaming `model_` (and by extension the `model()` accessor) in a future standalone PR?
</blockquote></details>

Fine with me. How about a TODO near the member and model() method as a reminder?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 957 at r3](https://reviewable.io/reviews/robotlocomotion/drake/9041#-LFdMqm82Rm-AnjXAogE:-LFdYKPpAIJOAp-O7M1_:b-iuekps) ([raw file](https://github.com/robotlocomotion/drake/blob/1956c90595321e7e3441bb9fa1ce4aec7d3eef91/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L957)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I don't think this was actually causing the assertion failure (I think it was because I left in the explicit -1 initializers for the geometry ports), but I've changed it to not use `push_back` anyway.
</blockquote></details>

Oh! My bad.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/9041#-:-LFdYyP-2odKuLChLh4l:b-hx8ib5)*
<!-- Sent from Reviewable.io -->
